src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -660,6 +538,25 @@ public static void Register(Type type, Type serializerType)
             }
         }
 
+        private static void AddAndExchangeDictionary<K, V>(ref IReadOnlyDictionary<K,V> currentDictionary, K keyToUpdate, V valueToUpdate, bool forceOverride)
+        {
+            V currentValue;
+            if (forceOverride || !currentDictionary.TryGetValue(keyToUpdate, out currentValue) || (currentValue == null))
+            {
+                if (addingComplete)
+                {
+                    var newDictionary = currentDictionary.ToDictionary(kv => kv.Key, kv => kv.Value);
+                    newDictionary[keyToUpdate] = valueToUpdate;
+                    System.Threading.Thread.MemoryBarrier();
+                    currentDictionary = new ReadOnlyDictionary<K,V>(newDictionary);
```

Comment:
I did not mean to use an actual ReadOnlyDictionary concrete instance, but just use IReadOnlyDictionary (which is already implemented by Dictionary). Not sure if the perf hit of wrapping in a ReadOnlyDictionary that prevents mutations is really necessary. Having said that, we probably don't mutate this very often, so it will not be in the hot path once the silo is warmed up. I'm not sure if there is a perf hit when reading using this wrapper class, so unless we test that, I'd say it's safer to just assign the Dictionary to the IReadOnlyDictionary field.